src/mxcp/endpoints/utils.py

```diff
@@ -0,0 +1,199 @@
+        endpoint_type: Type of endpoint ("tool", "resource", "prompt")
+        endpoint_file_path: Path to the endpoint YAML file
+        repo_root: Repository root path
+        
+    Returns:
+        The source code content
+        
+    Raises:
+        ValueError: If no source code found in endpoint definition
+    """
+    source = endpoint_dict[endpoint_type]["source"]
+    if "code" in source:
+        return source["code"]
+    elif "file" in source:
+        source_path = Path(source["file"])
+        if source_path.is_absolute():
+            full_path = repo_root / source_path.relative_to("/")
+        else:
+            full_path = endpoint_file_path.parent / source_path
+        return full_path.read_text()
+    else:
+        raise ValueError("No source code found in endpoint definition")
+
+
+def extract_source_info(source: Dict[str, Any]) -> Tuple[str, str]:
+    """Extract source code and determine if it's inline code or file reference.
+    
+    Args:
+        source: Source dictionary from endpoint definition
+        
+    Returns:
+        Tuple of (source_type, source_value) where:
+        - source_type: "code" or "file" 
+        - source_value: The actual code string or file path
+        
+    Raises:
+        ValueError: If no source code or file found
+    """
+    if "code" in source:
+        return ("code", source["code"])
+    elif "file" in source:
+        return ("file", source["file"])
+    else:
+        raise ValueError("No source code or file found in source definition")
+
+
+def detect_language_from_source(source: Dict[str, Any], file_path: Optional[str] = None) -> str:
+    """Detect programming language from source definition.
+    
+    Args:
+        source: Source dictionary from endpoint definition
+        file_path: Optional file path to use for extension-based detection
+        
+    Returns:
+        Language string ("python", "sql", etc.)
+    """
+    # Check if language is explicitly specified
+    if "language" in source:
+        return source["language"]
+    
+    # Try to infer from file extension
+    path_to_check = file_path or source.get("file")
+    if path_to_check:
+        if path_to_check.endswith((".py", ".python")):
+            return "python"
+        elif path_to_check.endswith((".sql", ".SQL")):
+            return "sql"
+    
+    # Default to SQL for backward compatibility
+    return "sql"
+
+
+def resolve_file_path(
+    file_path: str, 
+    endpoint_file_path: Path, 
+    repo_root: Path
+) -> Path:
+    """Resolve a relative file path to an absolute path.
+    
+    Args:
+        file_path: File path from source definition (may be relative)
+        endpoint_file_path: Path to the endpoint YAML file
+        repo_root: Repository root path
+        
+    Returns:
+        Resolved absolute path
+    """
+    source_path = Path(file_path)
+    if source_path.is_absolute():
+        return repo_root / source_path.relative_to("/")
```

Comment:
So if the path is _absolute_ we consider it as relative to the root? Like some kind of `chroot` would?

I can work with that, just thinking it's unusual.

Reply:
It must be absolute because the mxcp site/project has to be fully self-contained. You cannot reference things out of it.